File: server/services/commonConfigService.js
import { getCommonConfigJson, saveCommonConfigJson } from "../utils/utils.js";
import { COMMON_CONFIG_KEYS } from "../utils/constant.js";
import logger from "../utils/logger.js";

class CommonConfigService {
    constructor() {
        this.config = getCommonConfigJson();
        this.allowedKeys = Object.values(COMMON_CONFIG_KEYS);
    }

    /**
     * Validate if a key is allowed
     * @param {string} key - The key to validate
     * @returns {boolean} True if key is allowed, false otherwise
     */
    isValidKey(key) {
        return this.allowedKeys.includes(key);
    }

    /**
     * Throw error if key is invalid
     * @param {string} key - The key to validate
     * @throws {Error} If key is not allowed
     */
    validateKey(key) {
        if (!this.isValidKey(key)) {
            throw new Error(`Invalid config key: ${key}. Allowed keys are: ${this.allowedKeys.join(', ')}`);
        }
    }

    /**
     * Get all config data
     * @returns {Object} The entire config object
     */
    getAll() {
        return this.config;
    }

    /**
     * Get config value by key
     * @param {string} key - The key to retrieve
     * @returns {any} The value associated with the key, or undefined if not found
     */
    get(key) {
        this.validateKey(key);
        if (!(key in this.config)) {
            this.config[key] = null;
            saveCommonConfigJson(this.config);
        }
        return this.config[key];
    }

    /**
     * Update config value by key
     * @param {string} key - The key to update
     * @param {any} value - The new value
     * @param {boolean} [partial=false] - If true and both old and new values are objects, performs a partial update
     * @returns {boolean} True if update was successful, false otherwise
     */
    update(key, value, partial = false) {
        this.validateKey(key);
        try {
            if (partial && typeof this.config[key] === 'object' && typeof value === 'object') {
                this.config[key] = {
                    ...this.config[key],
                    ...value
                };
            } else {
                this.config[key] = value;
            }

            saveCommonConfigJson(this.config);
            return true;
        } catch (error) {
            logger.error('Error updating config:', error);
            return false;
        }
    }

    /**
     * Update multiple config values at once
     * @param {Object} updates - Object containing key-value pairs to update
     * @param {boolean} [partial=false] - If true, performs partial updates for object values
     * @returns {boolean} True if all updates were successful, false otherwise
     */
    updateMultiple(updates, partial = false) {
        try {
            Object.entries(updates).forEach(([key, value]) => {
                this.validateKey(key);
                if (partial && typeof this.config[key] === 'object' && typeof value === 'object') {
                    this.config[key] = {
                        ...this.config[key],
                        ...value
                    };
                } else {
                    this.config[key] = value;
                }
            });

            saveCommonConfigJson(this.config);
            return true;
        } catch (error) {
            logger.error('Error updating multiple configs:', error);
            return false;
        }
    }

    /**
     * Delete a config key
     * @param {string} key - The key to delete
     * @returns {boolean} True if deletion was successful, false otherwise
     */
    delete(key) {
        try {
            this.validateKey(key);
            delete this.config[key];
            saveCommonConfigJson(this.config);
            return true;
        } catch (error) {
            logger.error('Error deleting config:', error);
            return false;
        }
    }
}

export default new CommonConfigService();

File: app/services/common_config_service.py
from app.core.utils import get_common_config_json, save_common_config_json, get_default_playlist_json
from app.core.constants import COMMON_CONFIG_KEYS
from app.core import logger

class CommonConfigService:
    def __init__(self):
        self.config = get_common_config_json()
        self.allowed_keys = [
            COMMON_CONFIG_KEYS["defaultPlaylistGenre"]
        ]
        self.validations = self.setup_validations()
    
    def setup_validations(self):
        async def validate_playlist_value(value):
            if value == "all":
                return True
            
            playlists = get_default_playlist_json()
            genres = set(p.get('genre') for p in playlists)
            return value in genres
        
        return {
            COMMON_CONFIG_KEYS["defaultPlaylistGenre"]: {
                'validate': validate_playlist_value,
                'errorMessage': lambda value: f"Genre '{value}' does not exist in default playlists"
            }
        }
    
    def is_valid_key(self, key: str) -> bool:
        return key in self.allowed_keys
    
    async def validate_key_and_value(self, key: str, value):
        if not self.is_valid_key(key):
            raise Exception(f"Invalid config key: {key}. Allowed keys are: {', '.join(self.allowed_keys)}")
        
        if key in self.validations:
            validation = self.validations[key]
            is_valid = await validation['validate'](value)
            if not is_valid:
                raise Exception(validation['errorMessage'](value))
    
    def get_all(self):
        return self.config
    
    async def get(self, key: str = None):
        if key:
            return self.config.get(key)
        return self.config
    
    async def update(self, key: str, value, partial: bool = False):
        try:
            await self.validate_key_and_value(key, value)
            
            if partial and isinstance(self.config.get(key), dict) and isinstance(value, dict):
                self.config[key] = {
                    **self.config.get(key, {}),
                    **value
                }
            else:
                self.config[key] = value
            
            save_common_config_json(self.config)
            return True
        except Exception as error:
            logger.error('Error updating config', error=str(error))
            raise error
    
    async def update_multiple(self, updates: dict, partial: bool = False):
        try:
            for key, value in updates.items():
                await self.validate_key_and_value(key, value)
            
            for key, value in updates.items():
                if partial and isinstance(self.config.get(key), dict) and isinstance(value, dict):
                    self.config[key] = {
                        **self.config.get(key, {}),
                        **value
                    }
                else:
                    self.config[key] = value
            
            save_common_config_json(self.config)
            return True
        except Exception as error:
            logger.error('Error updating multiple configs', error=str(error))
            return False
    
    async def delete(self, key: str):
        try:
            if not self.is_valid_key(key):
                raise Exception(f"Invalid config key: {key}. Allowed keys are: {', '.join(self.allowed_keys)}")
            
            if key in self.config:
                del self.config[key]
            save_common_config_json(self.config)
            return True
        except Exception as error:
            logger.error('Error deleting config', error=str(error))
            return False

common_config_service = CommonConfigService()
